### ISSUE TITLE
How to write NaN in a shader?

### ISSUE THREAD

**Issue by unconed** [gpuweb/gpuweb#3749]
In the latest Chrome Canary, bitcasting a float to a NaN no longer works, i.e.:

```
  let NaN: f32 = bitcast<f32>(0xffffffffu);
```

Apparently this is intentional:
https://github.com/gpuweb/gpuweb/issues/3468

How am I supposed to return NaN from a shader now? This is a common way to knock out vertices from an otherwise continuous geometry stream.

**Comment by jrprice** [comment]
It's a little awkward, but you can define a function that does the equivalent without triggering errors from constant evaluation:
```
fn nan() -> f32 {
  let bits = 0xffffffffu;
  return bitcast<f32>(bits);
}
```

**Comment by unconed** [comment]
Based on the spec, it seems we're not supposed to assume NaN is actually supported though. Is there actual known hardware where NaNs don't work?

**Comment by ben-clayton** [comment]
Yes. And known backend compilers that can produce severely broken output with non-finites.

We will likely expose non-finites with a future extension, which will be available for drivers and hardware that correctly support it (i.e. pass thorough CTS).

**Comment by dneto0** [comment]
The relevant Vulkan feature bit is shaderSignedZeroInfNanPreserveFloat32 which was introduced by VK_KHR_shader_float_controls

It's not universally supported.
You can explore at https://vulkan.gpuinfo.org/listpropertiesextensions.php

**Comment by kainino0x** [comment]
> How am I supposed to return NaN from a shader now? This is a common way to knock out vertices from an otherwise continuous geometry stream.

Can you use a different sentinel value? Or does this technique rely on the hardware's handling of NaN vertices?
If the latter then the best you can do for now is James's suggestion.

**Comment by dneto0** [comment]
Ah, reopening as this is a documentation / developer advice issue.

**Comment by unconed** [comment]
> > How am I supposed to return NaN from a shader now? This is a common way to knock out vertices from an otherwise continuous geometry stream.
> 
> Can you use a different sentinel value? Or does this technique rely on the hardware's handling of NaN vertices? If the latter then the best you can do for now is James's suggestion.

There is no way to use a another sentinel value because the whole idea is that the entire triangle goes away if only one of the vertices is NaN. Doing this manually would require some coordination between the 3 vertex shader invocations.

I don't feel very confident in the proposed workaround because it basically violates the substitution principle: lifting a subexpression out into a variable should not turn invalid code into valid code or vice versa.

I've tried emitting vertices with w=0 which seems to work on Mac but may not work everywhere either.

**Comment by kainino0x** [comment]
> Ah, reopening as this is a documentation / developer advice issue.

Are we tracking this issue just for documentation or is it also tracking actual support for (the equivalent of Vulkan's) shaderSignedZeroInfNanPreserveFloat32/16? The request is raised by developers not-too-infrequently.

I also wonder how that feature is going to work with WGSL; would it just automatically enable SignedZeroInfNanPreserve in SPIR-V? Is there any performance benefit to allowing control over SignedZeroInfNanPreserve?

**Comment by dneto0** [comment]
#5109 proposes adding a diagnostic control to allow const-eval inf and nan.  If that goes in then you could write `const NaN = 0.0/0.0;`

**Comment by JD31** [comment]
Similar to #5316

**Comment by kainino0x** [comment]
I've collected some support info using [gpuinfo-vulkan-query](https://github.com/kainino0x/gpuinfo-vulkan-query):

[EDIT: analysis moved - go to <https://github.com/gpuweb/gpuweb/issues/2259#issuecomment-4139828366> for a more complete version]

**Comment by mwyrzykowski** [comment]
Based on https://github.com/gpuweb/gpuweb/issues/5316 I thought the WG resolved not to support this? Can this issue be duped to https://github.com/gpuweb/gpuweb/issues/5316?

**Comment by kainino0x** [comment]
Hm, you're right, this does seem very similar to that issue. But, I am unclear whether we resolved to (for now) plan not to add an (optional) feature for infinity/nan at all.

Maybe #2259 is the issue I wanted to post that on?
